tensorboard/data/server/reservoir.rs

```diff
@@ -120,19 +151,34 @@ impl<T> StageReservoir<T, ChaCha20Rng> {
 }
 
 impl<T, C: ReservoirControl> StageReservoir<T, C> {
+    /// Creates a new reservoir with the specified capacity and reservoir control.
+    ///
+    /// This function does not allocate. Reservoir capacity is allocated as records are offered.
+    pub fn with_control(capacity: usize, ctl: C) -> Self {
+        Self {
+            committed_steps: Vec::new(),
+            staged_items: Vec::new(),
+            capacity,
+            ctl,
+            seen: 0,
+        }
+    }
+
     /// Offers a record to the reservoir. The reservoir will always include the latest record.
     /// Other than the latest record, the records kept form a simple random sample of the stream
     /// (or at least approximately so in the case of preemptions).
     pub fn offer(&mut self, step: Step, v: T) {
+        if self.capacity == 0 {
```

Comment:
I’d be open to it, since the dead-read-store won’t cost much, but it
gets weird once you add preemptions:

```rust
if self.capacity == 0 {
    return;
}
self.preempt(step);
self.seen += 1;
```

It’s important to preempt before incrementing `seen`. And it seems good
to return on zero-capacity before preempting—otherwise, `preempt` has to
be more careful to work in the zero-capacity case.

Since `seen` is not exposed and reservoir capacity can’t be changed
after construction, this should be unobservable—and I was careful to
test the full public API in `test_empty` for this reason. So I think
I’ll keep it as is, though I don’t feel super strongly about it.

Added a clarifying note to the docs.